src/tactic/linarith.lean

```diff
@@ -98,6 +98,103 @@ by rw [←h3, mul_assoc, mul_div_comm, h2, ←mul_assoc, h1, mul_comm, one_mul]
 
 end lemmas
 
+/--
+A linear expression is a list of pairs of variable indices and coefficients.
+
+Some functions on `linexp` assume that `n : ℕ` occurs at most once as the first element of a pair,
+and that the list is sorted in decreasing order of the first argument.
+This is not enforced by the type but the operations here preserve it.
+-/
+@[reducible]
+def linexp : Type := list (ℕ × ℤ)
+end linarith
+
+/--
+A map `ℕ → ℤ` is converted to `list (ℕ × ℤ)` in the obvious way.
+This list is sorted in decreasing order of the first argument.
+-/
+meta def native.rb_map.to_linexp (m : rb_map ℕ ℤ) : linarith.linexp :=
+m.to_list
+
+namespace linarith
+namespace linexp
+
+/--
+Add two `linexp`s together componentwise.
+Preserves sorting and uniqueness of the first argument.
+-/
+def add : linexp → linexp → linexp
```

Comment:
```suggestion
meta def add : linexp → linexp → linexp
```

For performance reasons, surprisingly.  (This function is otherwise defined using well-founded recursion, which generates horrible meta code.)

Reply:
Crap, really? I thought the meta code that gets generated was the same whether or not there was a `meta` flag.

Reply:
Yes, you can check the generated code using `set_option trace.compiler.optimize_bytecode true`.  The well-founded recursion is implemented by packing the arguments into a tuple, and you can see this in the bytecode by the auxiliary definitions containing the name `_packed`.